pkg/protocol/engine/upgrade/orchestrator.go

```diff
@@ -0,0 +1,329 @@
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/blocks"
+	"github.com/iotaledger/iota-core/pkg/protocol/sybilprotection/seatmanager"
+	"github.com/iotaledger/iota-core/pkg/storage/prunable"
+	"github.com/iotaledger/iota-core/pkg/votes"
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+// Orchestrator is a component that is in charge of protocol upgrades by signaling the upgrade, gathering
+// consensus about the upgrade and finally activating the upgrade if the activation thresholds are reached.
+//
+// Why is it called Orchestrator? Read on...
+//
+// User
+// how would you name a software component that is in charge of protocol upgrades by signaling the upgrade, gathering
+// consensus about the upgrade and then activating the upgrade.
+//
+// ChatGPT
+// One possible name for such a software component could be the "Upgrade Manager." This name conveys its responsibility
+// of managing protocol upgrades within a system or network. The Upgrade Manager would handle tasks like signaling,
+// consensus gathering, and activation of upgrades, ensuring smooth transitions and maintaining the integrity of the protocol.
+//
+// User
+// do you have a less boring name? everything else is also called manager
+//
+// ChatGPT
+// Certainly! If you're looking for a more creative and less generic name, you could consider calling it the "Protocol Phoenix."
+// This name captures the essence of the component's role in orchestrating protocol upgrades, as the mythical phoenix
+// symbolizes rebirth and renewal. It signifies the component's ability to facilitate the evolution and transformation
+// of the protocol, rising from the ashes of the previous version.
+//
+// User
+// maybe a more technical and less mythical term
+//
+// ChatGPT
+// Certainly! In that case, you could consider naming the software component the "Upgrade Orchestrator."
+// This name emphasizes its role in orchestrating protocol upgrades, managing the various stages of the process, and
+// ensuring a smooth transition. The term "orchestrator" suggests a conductor-like function, coordinating the upgrade
+// activities and ensuring consensus before activating the changes.
+type Orchestrator struct {
+	evictionMutex syncutils.RWMutex
+	errorHandler  func(error)
+
+	latestSignals           *memstorage.IndexedStorage[iotago.SlotIndex, account.SeatIndex, *prunable.SignaledBlock]
+	upgradeSignalsFunc      func(slot iotago.SlotIndex) *prunable.UpgradeSignals
+	permanentUpgradeSignals *kvstore.TypedStore[iotago.EpochIndex, VersionAndHash]
+
+	setProtocolParametersEpochMappingFunc func(iotago.Version, iotago.EpochIndex) error
+	protocolParametersAndVersionsHashFunc func(iotago.SlotIndex) (iotago.Identifier, error)
+	epochForVersionFunc                   func(iotago.Version) (iotago.EpochIndex, bool)
+
+	apiProvider api.Provider
+	seatManager seatmanager.SeatManager
+
+	module.Module
+}
+
+func NewProvider(opts ...options.Option[Orchestrator]) module.Provider[*engine.Engine, *Orchestrator] {
```

Comment:
I am missing the way to provide the new set of version+protocolparams from the config